{
  "cve_id": "CVE-2022-35934",
  "cve_desc": "TensorFlow is an open source platform for machine learning. The implementation of tf.reshape op in TensorFlow is vulnerable to a denial of service via CHECK-failure (assertion failure) caused by overflowing the number of elements in a tensor. This issue has been patched in GitHub commit 61f0f9b94df8c0411f0ad0ecc2fec2d3f3c33555. The fix will be included in TensorFlow 2.10.0. We will also cherrypick this commit on TensorFlow 2.9.1, TensorFlow 2.8.1, and TensorFlow 2.7.2, as these are also affected and still in supported range. There are no known workarounds for this issue.",
  "repo": "tensorflow/tensorflow",
  "patch_hash": "61f0f9b94df8c0411f0ad0ecc2fec2d3f3c33555",
  "patch_info": {
    "commit_hash": "61f0f9b94df8c0411f0ad0ecc2fec2d3f3c33555",
    "repo": "tensorflow/tensorflow",
    "commit_url": "https://github.com/tensorflow/tensorflow/commit/61f0f9b94df8c0411f0ad0ecc2fec2d3f3c33555",
    "files": [
      "tensorflow/core/kernels/reshape_op.h",
      "tensorflow/python/kernel_tests/array_ops/array_ops_test.py"
    ],
    "message": "Fix failed check in tf.reshape.\n\nPassing in too many dimensions causes an assertion\nfailure and a crash. Check input as a precondition and\nfail gracefully.\n\nPiperOrigin-RevId: 449893553",
    "before_after_code_files": [
      "tensorflow/core/kernels/reshape_op.h||tensorflow/core/kernels/reshape_op.h",
      "tensorflow/python/kernel_tests/array_ops/array_ops_test.py||tensorflow/python/kernel_tests/array_ops/array_ops_test.py"
    ]
  },
  "patch_diff": {
    "tensorflow/core/kernels/reshape_op.h||tensorflow/core/kernels/reshape_op.h": [
      "File: tensorflow/core/kernels/reshape_op.h -> tensorflow/core/kernels/reshape_op.h",
      "--- Hunk 1 ---",
      "[Context before]",
      "45:          TensorShapeUtils::IsScalar(sizes.shape())),",
      "46:         errors::InvalidArgument(\"sizes input must be 1-D, not \",",
      "47:                                 sizes.shape().DebugString()));",
      "",
      "[Removed Lines]",
      "[None]",
      "",
      "[Added Lines]",
      "48:     OP_REQUIRES(",
      "49:         context, sizes.NumElements() < TensorShape::MaxDimensions(),",
      "50:         errors::InvalidArgument(\"too many dimensions: must be < \",",
      "51:                                 TensorShape::MaxDimensions(), \", but received \",",
      "52:                                 sizes.NumElements()));",
      "",
      "---------------"
    ],
    "tensorflow/python/kernel_tests/array_ops/array_ops_test.py||tensorflow/python/kernel_tests/array_ops/array_ops_test.py": [
      "File: tensorflow/python/kernel_tests/array_ops/array_ops_test.py -> tensorflow/python/kernel_tests/array_ops/array_ops_test.py",
      "--- Hunk 1 ---",
      "[Context before]",
      "351:                                 \"must be a tensor with a single value\"):",
      "352:       array_ops.expand_dims(1, axis=[0, 1])",
      "355: @test_util.with_eager_op_as_function",
      "356: class ReverseV2Test(test_util.TensorFlowTestCase):",
      "",
      "[Removed Lines]",
      "[None]",
      "",
      "[Added Lines]",
      "354:   def testReshapeWithManyDims(self):",
      "355:     with self.assertRaisesRegex(errors.InvalidArgumentError,",
      "356:                                 \"too many dimensions\"):",
      "357:       self.evaluate(",
      "358:           array_ops.reshape(",
      "359:               tensor=[[1]],",
      "360:               shape=constant_op.constant([1 for i in range(254)],",
      "361:                                          dtype=dtypes.int64)))",
      "",
      "---------------"
    ]
  },
  "candidates": [
    {
      "candidate_hash": "7028c67fcea549ada8783f540ef63d54a61d2e76",
      "candidate_info": {
        "commit_hash": "7028c67fcea549ada8783f540ef63d54a61d2e76",
        "repo": "tensorflow/tensorflow",
        "commit_url": "https://github.com/tensorflow/tensorflow/commit/7028c67fcea549ada8783f540ef63d54a61d2e76",
        "files": [
          "tensorflow/core/kernels/reshape_op.h",
          "tensorflow/python/kernel_tests/array_ops/array_ops_test.py"
        ],
        "message": "Fix failed check in tf.reshape.\n\nPassing in too many dimensions causes an assertion\nfailure and a crash. Check input as a precondition and\nfail gracefully.\n\nPiperOrigin-RevId: 449893553",
        "before_after_code_files": [
          "tensorflow/core/kernels/reshape_op.h||tensorflow/core/kernels/reshape_op.h",
          "tensorflow/python/kernel_tests/array_ops/array_ops_test.py||tensorflow/python/kernel_tests/array_ops/array_ops_test.py"
        ]
      },
      "candidate_patch_features": {
        "candidate_earlier_than_patch": 0,
        "diff_branch_same_aad": 1,
        "olp_code_files": {
          "patch": [
            "tensorflow/core/kernels/reshape_op.h||tensorflow/core/kernels/reshape_op.h",
            "tensorflow/python/kernel_tests/array_ops/array_ops_test.py||tensorflow/python/kernel_tests/array_ops/array_ops_test.py"
          ],
          "candidate": [
            "tensorflow/core/kernels/reshape_op.h||tensorflow/core/kernels/reshape_op.h",
            "tensorflow/python/kernel_tests/array_ops/array_ops_test.py||tensorflow/python/kernel_tests/array_ops/array_ops_test.py"
          ]
        }
      },
      "candidate_diff": {
        "tensorflow/core/kernels/reshape_op.h||tensorflow/core/kernels/reshape_op.h": [
          "File: tensorflow/core/kernels/reshape_op.h -> tensorflow/core/kernels/reshape_op.h",
          "--- Hunk 1 ---",
          "[Context before]",
          "45:          TensorShapeUtils::IsScalar(sizes.shape())),",
          "46:         errors::InvalidArgument(\"sizes input must be 1-D, not \",",
          "47:                                 sizes.shape().DebugString()));",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "48:     OP_REQUIRES(",
          "49:         context, sizes.NumElements() < TensorShape::MaxDimensions(),",
          "50:         errors::InvalidArgument(\"too many dimensions: must be < \",",
          "51:                                 TensorShape::MaxDimensions(), \", but received \",",
          "52:                                 sizes.NumElements()));",
          "",
          "---------------"
        ],
        "tensorflow/python/kernel_tests/array_ops/array_ops_test.py||tensorflow/python/kernel_tests/array_ops/array_ops_test.py": [
          "File: tensorflow/python/kernel_tests/array_ops/array_ops_test.py -> tensorflow/python/kernel_tests/array_ops/array_ops_test.py",
          "--- Hunk 1 ---",
          "[Context before]",
          "351:                                 \"must be a tensor with a single value\"):",
          "352:       array_ops.expand_dims(1, axis=[0, 1])",
          "355: @test_util.with_eager_op_as_function",
          "356: class ReverseV2Test(test_util.TensorFlowTestCase):",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "354:   def testReshapeWithManyDims(self):",
          "355:     with self.assertRaisesRegex(errors.InvalidArgumentError,",
          "356:                                 \"too many dimensions\"):",
          "357:       self.evaluate(",
          "358:           array_ops.reshape(",
          "359:               tensor=[[1]],",
          "360:               shape=constant_op.constant([1 for i in range(254)],",
          "361:                                          dtype=dtypes.int64)))",
          "",
          "---------------"
        ]
      }
    },
    {
      "candidate_hash": "f983c1129c61079491d34e7085ee2c446d1188fb",
      "candidate_info": {
        "commit_hash": "f983c1129c61079491d34e7085ee2c446d1188fb",
        "repo": "tensorflow/tensorflow",
        "commit_url": "https://github.com/tensorflow/tensorflow/commit/f983c1129c61079491d34e7085ee2c446d1188fb",
        "files": [
          "tensorflow/core/kernels/reshape_op.h",
          "tensorflow/python/kernel_tests/array_ops_test.py"
        ],
        "message": "Fix failed check in tf.reshape.\n\nPassing in too many dimensions causes an assertion\nfailure and a crash. Check input as a precondition and\nfail gracefully.\n\nPiperOrigin-RevId: 449893553",
        "before_after_code_files": [
          "tensorflow/core/kernels/reshape_op.h||tensorflow/core/kernels/reshape_op.h",
          "tensorflow/python/kernel_tests/array_ops_test.py||tensorflow/python/kernel_tests/array_ops_test.py"
        ]
      },
      "candidate_patch_features": {
        "candidate_earlier_than_patch": 0,
        "diff_branch_same_aad": 1,
        "olp_code_files": {
          "patch": [
            "tensorflow/core/kernels/reshape_op.h||tensorflow/core/kernels/reshape_op.h"
          ],
          "candidate": [
            "tensorflow/core/kernels/reshape_op.h||tensorflow/core/kernels/reshape_op.h"
          ]
        }
      },
      "candidate_diff": {
        "tensorflow/core/kernels/reshape_op.h||tensorflow/core/kernels/reshape_op.h": [
          "File: tensorflow/core/kernels/reshape_op.h -> tensorflow/core/kernels/reshape_op.h",
          "--- Hunk 1 ---",
          "[Context before]",
          "45:          TensorShapeUtils::IsScalar(sizes.shape())),",
          "46:         errors::InvalidArgument(\"sizes input must be 1-D, not \",",
          "47:                                 sizes.shape().DebugString()));",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "48:     OP_REQUIRES(",
          "49:         context, sizes.NumElements() < TensorShape::MaxDimensions(),",
          "50:         errors::InvalidArgument(\"too many dimensions: must be < \",",
          "51:                                 TensorShape::MaxDimensions(), \", but received \",",
          "52:                                 sizes.NumElements()));",
          "",
          "---------------"
        ],
        "tensorflow/python/kernel_tests/array_ops_test.py||tensorflow/python/kernel_tests/array_ops_test.py": [
          "File: tensorflow/python/kernel_tests/array_ops_test.py -> tensorflow/python/kernel_tests/array_ops_test.py",
          "--- Hunk 1 ---",
          "[Context before]",
          "353:                                 \"must be a tensor with a single value\"):",
          "354:       array_ops.expand_dims(1, axis=[0, 1])",
          "357: class ReverseV2Test(test_util.TensorFlowTestCase):",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "356:   def testReshapeWithManyDims(self):",
          "357:     with self.assertRaisesRegex(errors.InvalidArgumentError,",
          "358:                                 \"too many dimensions\"):",
          "359:       self.evaluate(",
          "360:           array_ops.reshape(",
          "361:               tensor=[[1]],",
          "362:               shape=constant_op.constant([1 for i in range(254)],",
          "363:                                          dtype=dtypes.int64)))",
          "",
          "---------------"
        ]
      }
    },
    {
      "candidate_hash": "f5e9a48a982484ff027e8b19bbc8d46aa7594f8d",
      "candidate_info": {
        "commit_hash": "f5e9a48a982484ff027e8b19bbc8d46aa7594f8d",
        "repo": "tensorflow/tensorflow",
        "commit_url": "https://github.com/tensorflow/tensorflow/commit/f5e9a48a982484ff027e8b19bbc8d46aa7594f8d",
        "files": [
          "tensorflow/core/kernels/reshape_op.h",
          "tensorflow/python/kernel_tests/array_ops/array_ops_test.py"
        ],
        "message": "Fix failed check in tf.reshape.\n\nPassing in too many dimensions causes an assertion\nfailure and a crash. Check input as a precondition and\nfail gracefully.\n\nPiperOrigin-RevId: 449893553",
        "before_after_code_files": [
          "tensorflow/core/kernels/reshape_op.h||tensorflow/core/kernels/reshape_op.h",
          "tensorflow/python/kernel_tests/array_ops/array_ops_test.py||tensorflow/python/kernel_tests/array_ops/array_ops_test.py"
        ]
      },
      "candidate_patch_features": {
        "candidate_earlier_than_patch": 0,
        "diff_branch_same_aad": 1,
        "olp_code_files": {
          "patch": [
            "tensorflow/core/kernels/reshape_op.h||tensorflow/core/kernels/reshape_op.h",
            "tensorflow/python/kernel_tests/array_ops/array_ops_test.py||tensorflow/python/kernel_tests/array_ops/array_ops_test.py"
          ],
          "candidate": [
            "tensorflow/core/kernels/reshape_op.h||tensorflow/core/kernels/reshape_op.h",
            "tensorflow/python/kernel_tests/array_ops/array_ops_test.py||tensorflow/python/kernel_tests/array_ops/array_ops_test.py"
          ]
        }
      },
      "candidate_diff": {
        "tensorflow/core/kernels/reshape_op.h||tensorflow/core/kernels/reshape_op.h": [
          "File: tensorflow/core/kernels/reshape_op.h -> tensorflow/core/kernels/reshape_op.h",
          "--- Hunk 1 ---",
          "[Context before]",
          "45:          TensorShapeUtils::IsScalar(sizes.shape())),",
          "46:         errors::InvalidArgument(\"sizes input must be 1-D, not \",",
          "47:                                 sizes.shape().DebugString()));",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "48:     OP_REQUIRES(",
          "49:         context, sizes.NumElements() < TensorShape::MaxDimensions(),",
          "50:         errors::InvalidArgument(\"too many dimensions: must be < \",",
          "51:                                 TensorShape::MaxDimensions(), \", but received \",",
          "52:                                 sizes.NumElements()));",
          "",
          "---------------"
        ],
        "tensorflow/python/kernel_tests/array_ops/array_ops_test.py||tensorflow/python/kernel_tests/array_ops/array_ops_test.py": [
          "File: tensorflow/python/kernel_tests/array_ops/array_ops_test.py -> tensorflow/python/kernel_tests/array_ops/array_ops_test.py",
          "--- Hunk 1 ---",
          "[Context before]",
          "351:                                 \"must be a tensor with a single value\"):",
          "352:       array_ops.expand_dims(1, axis=[0, 1])",
          "355: @test_util.with_eager_op_as_function",
          "356: class ReverseV2Test(test_util.TensorFlowTestCase):",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "354:   def testReshapeWithManyDims(self):",
          "355:     with self.assertRaisesRegex(errors.InvalidArgumentError,",
          "356:                                 \"too many dimensions\"):",
          "357:       self.evaluate(",
          "358:           array_ops.reshape(",
          "359:               tensor=[[1]],",
          "360:               shape=constant_op.constant([1 for i in range(254)],",
          "361:                                          dtype=dtypes.int64)))",
          "",
          "---------------"
        ]
      }
    }
  ]
}